README.md
# tilelive-http

I am an HTTP source for [tilelive](https://github.com/mapbox/tilelive.js).

## Usage

```javascript
var tilelive = require("@mapbox/tilelive");
require("tilelive-http")(tilelive);
require("@mapbox/mbtiles").registerProtocols(tilelive);

var template = "http://tile.stamen.com/watercolor/{z}/{x}/{y}.jpg";

var scheme = tilelive.Scheme.create("scanline", {
  minzoom: 10,
  maxzoom: 11,
  bbox: [-118.9448, 32.8007, -117.6462, 34.8233]
});

var copyTask = new tilelive.CopyTask(template, "mbtiles://./watercolor-la.mbtiles", scheme);
copyTask.formats = ["tile"];

copyTask.start(function(err) {
  if (err) {
    throw err;
  }
});

copyTask.on("progress", console.log);
copyTask.on("finished", function() {
  console.log("Done!");
});
```

## Environment Variables

* `TILELIVE_USER_AGENT` - `User-Agent` header when making upstream requests

## Gotchas

Vector tiles are gzipped at rest. They are served with Content-Encoding: gzip for transport, but by default `tilelive-http` does not automatically decompress them. See [mojodna/tilelive-http#13](https://github.com/mojodna/tilelive-http/pull/13) and [mapbox/vector-tile-spec#27 (comment)](https://github.com/mapbox/vector-tile-spec/issues/27#issuecomment-66670565) for more details.


CHANGELOG.md
# Changes

## v0.14.0 - 11/24/17

* Track `Age`, `Date`, and `Expires` HTTP headers to facilitate invalidation.

## v0.13.0 - 9/29/17

* Add `Accept-encoding: gzip, deflate` as the default with
  `option.acceptEncoding` to override it. This facilitates accessing correctly
  configured MVT services like maps.wikimedia.org

## v0.12.2 - 6/14/17

* Fix dependency `require`s

## v0.12.1 - 6/13/17

* Re-released with fixes from v0.11.3

## v0.12.0 - 6/13/17

* Update dependencies

## v0.11.3 - 7/24/16

* Revert explicit request for `gzip`ed content, as MVT PBFs are intended to be
  compressed at rest.

## v0.11.2 - 7/14/16

* Request `gzip` `Content-Encoding` to trigger appropriate response headers

## v0.11.1 - 7/10/16

* Typo fixes

## v0.11.0 - 7/10/16

* Allow `info` to be specified on the `uri` object
* Limit requests to bounds / zoom range specified in `info`
* Respect `info.autoscale` (defaults to `true`)
* Update dependencies
* Drop `options.concurrency` now that `maxSockets` defaults to `Infinity` (Node-0.12+)

## v0.10.1 - 7/6/16

* Pass `cache-control` headers through

## v0.10.0 - 2/22/16

* Support setting `User-Agent` header via `options.userAgent` /
  `TILELIVE_USER_AGENT`.

## v0.9.0 - 1/7/16

* Use `debug` for debugging. Set `DEBUG=tilelive-http` in the environment to
  enable.
* Reduce retry factor for a total waiting time of 5 minutes per tile.

## v0.8.0 - 6/6/15

* Version-specific workaround for joyent/node#25636 / nodejs/io.js#2113
  (breaking change in `url.parse` behavior)

## v0.7.0 - 5/9/15

* Make retries optional (`options.retry`, defaulting to `false`)

## v0.6.1 - 4/30/15

* Treat 403 responses as fatal errors
* Include problematic URLs in error messages
* Set `maxSockets` according to `options.concurrency`
* Revert to `retry` defaults; better behavior with problematic sources

## v0.6.0 - 4/14/15

* Retry failed requests (internal errors or 5xx responses)
* Downcase placeholder characters (for ModestMaps URL template compatibility)

## v0.5.0 - 10/15/14

* Relax `tilejson` dependency
* Report `autoscale: true`

## v0.4.0 - 7/25/14

* `?scale` awareness; generates URLs with `@<scale>x`
* Relax `tilelive` dependency
* Add quadkey interpolation for `{q}`

## v0.3.0 - 7/5/14

* Update `tilejson` dependency to `^0.8.0'.

## v0.2.3 - 5/11/14

* Depend on `tilejson@^0.6.4`, as upstream has picked up the protocol change.

## v0.2.2 - 5/8/14

* Point to the version of `tilejson` that registers for `tilejson+<proto>`

## v0.2.1 - 5/7/14

* Add `tilejson` as a peer dependency

## v0.2.0 - 5/7/14

* Disambiguate between URL templates and TileJSON sources (deferring to
  [tilejson](https://github.com/mapbox/node-tilejson) for the latter)
* Switch to factory construction to make `tilelive` available without directly
  depending on it

## v0.1.2 - 3/17/14

* Return an `Error` on 404s to match `tilelive.js` expectations (@gravitystorm)
* Register for `tilejson+https:`

## v0.1.1 - 12/19/13

* Implement `getInfo()` with sensible defaults
* Correct `copyTask.formats` in the docs and examples (@vsivsi)
* Initial released version


index.js
"use strict";

var http = require("http"),
    url = require("url"),
    util = require("util");

var _debug = require("debug"),
    mercator = new (require("@mapbox/sphericalmercator"))(),
    request = require("request"),
    retry = require("retry"),
    semver = require("semver");

var meta = require("./package.json"),
    NAME = meta.name,
    VERSION = meta.version,
    debug = _debug(NAME);

var quadKey = function(zoom, x, y) {
  var key = "";

  for (var i = zoom; i > 0; i--) {
    var digit = "0";
    var mask = 1 << (i - 1);
    if ((x & mask) !== 0) {
      digit++;
    }
    if ((y & mask) !== 0) {
      digit++;
      digit++;
    }
    key += digit;
  }

  return key;
};

module.exports = function(tilelive, options) {
  options = options || {};
  options.retry = "retry" in options ? options.retry : false;
  options.userAgent = options.userAgent || process.env.TILELIVE_USER_AGENT || [NAME, VERSION].join("/");
  options.acceptEncoding = options.acceptEncoding || 'gzip, deflate';

  var headers = {
        "User-Agent": options.userAgent,
        "Accept-encoding": options.acceptEncoding
      },
      retryOptions = {
        factor: 1.71023
      };

  if (!options.retry) {
    retryOptions.retries = 0;
  }

  var fetch = function(uri, callback) {
    var operation = retry.operation(retryOptions);

    return operation.attempt(function(currentAttempt) {
      return request.get({
        uri: uri,
        encoding: null,
        headers: headers,
        timeout: 30e3
      }, function(err, rsp, body) {
        if (operation.retry(err)) {
          debug("Failed %s after %d attempts:", url.format(uri), currentAttempt, err);

          // retrying, callback will eventually be called
          return;
        }

        if (err) {
          return callback(operation.mainError());
        }

        switch (rsp.statusCode) {
        case 200:
        case 403:
        case 404:
          return callback(null, rsp, body);

        default:
          err = new Error(util.format("Upstream error: %s returned %d", uri, rsp.statusCode));

          if (rsp.statusCode.toString().slice(0, 1) !== "5") {
            return callback(err);
          }

          if (!operation.retry(err)) {
            return callback(operation.mainError(), rsp, body);
          } else {
            debug("Failed %s after %d attempts:", url.format(uri), currentAttempt, err);
          }
        }
      });
    });
  };

  var HttpSource = function(uri, callback) {
    if (semver.satisfies(process.version, ">=0.11.0")) {
      // Node 0.12 changes the behavior of url.parse such that components are
      // url-encoded
      uri.hash = uri.hash && decodeURIComponent(uri.hash);
      uri.pathname = decodeURIComponent(uri.pathname);
      uri.path = decodeURIComponent(uri.path);
      uri.href = decodeURIComponent(uri.href);
    }

    this.source = url.format(uri).replace(/(\{\w\})/g, function(x) {
      return x.toLowerCase();
    });

    if (!(this.source.match(/{z}/) &&
        this.source.match(/{x}/) &&
        this.source.match(/{y}/)) &&
       !this.source.match(/{q}/)) {
      console.log("Coordinate placeholders missing; assuming %s is a TileJSON endpoint (tilejson+).", this.source);

      return tilelive.load("tilejson+" + this.source, callback);
    }

    this.scale = uri.query.scale || 1;
    this.tileSize = (uri.query.tileSize | 0) || 256;

    // save ourselves some math if we don't need to generate a quad key
    if (this.source.match(/{q}/)) {
      this.quadKey = quadKey;
    } else {
      this.quadKey = function() {};
    }

    // abuse the URI object by looking for .info directly on it
    this.info = uri.info || {};

    this.info.autoscale = "autoscale" in this.info ? this.info.autoscale : true;
    this.info.bounds = this.info.bounds || [-180, -85.0511, 180, 85.0511];
    this.info.minzoom = "minzoom" in this.info ? this.info.minzoom : 0;
    this.info.maxzoom = "maxzoom" in this.info ? this.info.maxzoom : Infinity;

    return callback(null, this);
  };

  HttpSource.prototype.getTile = function(z, x, y, callback) {
    var tileUrl = this.source
      .replace(/{z}/i, z)
      .replace(/{x}/i, x)
      .replace(/{y}/i, y)
      .replace(/{q}/i, this.quadKey(z, x, y));

    if (this.scale > 1 && this.info.autoscale) {
      // replace the last "." with "@<scale>x."
      tileUrl = tileUrl.replace(/\.(?!.*\.)/, "@" + this.scale + "x.");
    }

    if (z < this.info.minzoom || z > this.info.maxzoom) {
      return callback(new Error("Tile does not exist"));
    }

    var xyz = mercator.xyz(this.info.bounds, z);

    if (x < xyz.minX ||
        x > xyz.maxX ||
        y < xyz.minY ||
        y > xyz.maxY) {
      return callback(new Error("Tile does not exist"));
    }

    return fetch(tileUrl, function(err, rsp, body) {
      if (err) {
        return callback(err);
      }

      switch (rsp.statusCode) {
      case 200:
        var rspHeaders = ["age", "cache-control", "content-type", "content-md5", "content-encoding", "date", "expires"].reduce(function(obj, key) {
          if (rsp.headers[key]) {
            obj[key] = rsp.headers[key];
          }

          return obj;
        }, {});

        return callback(null, body, rspHeaders);

      case 404:
        return callback(new Error("Tile does not exist"));

      default:
        return callback(new Error("Upstream error: " + rsp.statusCode));
      }
    });
  };

  HttpSource.prototype.getInfo = function(callback) {
    return setImmediate(callback, null, this.info);
  };

  HttpSource.prototype.close = function(callback) {
    callback = callback || function() {};

    return callback();
  };

  HttpSource.registerProtocols = function(_tilelive) {
    _tilelive.protocols["http:"] = HttpSource;
    _tilelive.protocols["https:"] = HttpSource;
  };

  HttpSource.registerProtocols(tilelive);

  return HttpSource;
};
